### PR TITLE
Task #209083: changes in cron job

### PR DESCRIPTION
**Pull request by rushi-tekdi**
-added 2 columns in table lms_test_tracking
cron_last_processed_at- timestamp with time zone, nullable, default: now()
attendance_count- numeric, nullable
-added last processed logic for cron job
-maintain attedance count status ion cron job
=added new added columns in get response of graphql query

# I have ensured that following `Pull Request Checklist` is taken care of before sending this PR

* [ ] Code is formatted as per format decided
* [ ] Updated acceptance criteria in tracker
* [ ] Updated test cases in test-cases-tracker
* [ ] Updated new API endpoint if any in common postman collection
* [ ] Updated DB changes in db-tracker if any

## Summary by CodeRabbit

- **New Features**
	- Improved attendance data retrieval with added validation to ensure consistent user experience.
	- Enhanced certificate generation process with new filters and tracking updates for better accuracy and monitoring.

- **Bug Fixes**
	- Ensured user names are correctly fetched, with a fallback to an empty array to prevent potential errors.

- **Refactor**
	- Streamlined Learning Management System (LMS) service data handling with updated fields and logic to better reflect user progress and achievements.

- **Chores**
	- Implemented additional logging for key processes to aid in troubleshooting and maintenance.

### CODE REVIEW

File: src/src/attendances/attendances.core.service.ts [hunk 1]
```diff
@@ -241,10 +241,17 @@ export class AttendancesCoreService {
...
 		try {
-			const data_list = (
-				await this.hasuraServiceFromServices.getData({ query })
-			)?.data?.attendance;
-			return data_list;
+			const result_response =
+				await this.hasuraServiceFromServices.getData({ query });
+			console.log('result_response', result_response);
+			const data_list = result_response?.data?.attendance;
+			//console.log('data_list cunt------>>>>>', data_list.length);
```
Comment (manojLondhe) [inline src/src/attendances/attendances.core.service.ts:248]: remove console logs pls

File: src/src/cron/faFaceIndexing.cron.ts [hunk 2]
```diff
@@ -200,7 +200,7 @@ export class FaFaceIndexingCron {
 		// We need to skip processing records wch were processed in past X hours
 		let dt = new Date();
 		let filterTimestamp = new Date(
-			dt.setHours(dt.getHours() - 4),
+			dt.setHours(dt.getHours() - 2),
```
Comment (manojLondhe) [inline src/src/cron/faFaceIndexing.cron.ts:203]: Get this `2` from config / env

File: src/src/attendances/attendances.core.service.ts [hunk 3]
```diff
@@ -241,10 +241,17 @@ export class AttendancesCoreService {
 				  }
 			  }`;
 		try {
-			const data_list = (
-				await this.hasuraServiceFromServices.getData({ query })
-			)?.data?.attendance;
-			return data_list;
+			const result_response =
+				await this.hasuraServiceFromServices.getData({ query });
+			console.log('result_response', result_response);
```
Comment (manojLondhe) [inline src/src/attendances/attendances.core.service.ts:246]: remove console logs pls

File: src/src/cron/prepareCertificateHtml.cron.ts [hunk 4]
```diff
@@ -35,6 +35,7 @@ export class PrepareCertificateHtmlCron {
 				),
 			),
 		);
+		console.log('userForIssueCertificate list', userForIssueCertificate);
```
Comment (manojLondhe) [inline src/src/cron/prepareCertificateHtml.cron.ts:38]: remove console logs pls

File: src/src/cron/prepareCertificateHtml.cron.ts [hunk 5]
```diff
@@ -75,6 +76,7 @@ export class PrepareCertificateHtmlCron {
 						context,
 						context_id,
 					);
+				console.log('usrAttendanceList list', usrAttendanceList);
```
Comment (manojLondhe) [inline src/src/cron/prepareCertificateHtml.cron.ts:79]: remove console logs pls

File: src/src/cron/faFaceIndexing.cron.ts (hunk 2 above)
Comment (manojLondhe) [reply src/src/cron/faFaceIndexing.cron.ts:203]: what is 2 for?

File: src/src/cron/prepareCertificateHtml.cron.ts [hunk 6]
```diff
@@ -191,22 +193,47 @@ export class PrepareCertificateHtmlCron {
...
+					testTrackingUpdateData,
 				);
 			}
 		}
 	}
 	async fetchTestTrackingData(limit: number) {
+		// We need to skip processing records wch were processed in past X hours
+		let dt = new Date();
+		let filterTimestamp = new Date(
+			dt.setHours(dt.getHours() - 4),
```
Comment (manojLondhe) [inline src/src/cron/prepareCertificateHtml.cron.ts:218]: Get this 4 from config / env

File: src/src/attendances/attendances.core.service.ts (hunk 1 above)
Comment (rushi-tekdi) [reply src/src/attendances/attendances.core.service.ts:248]: @manojLondhe  console log on dev is required for testing and track cron job flow. after testing complete on dev server then i will remove all logs and sent PR again

File: src/src/cron/faFaceIndexing.cron.ts (hunk 2 above)
Comment (rushi-tekdi) [reply src/src/cron/faFaceIndexing.cron.ts:203]: 2 is hours limit for last_processed. just like 4 we used in attedance face rekognition.
after testing done on dev I will move it in env varibale

File: src/src/attendances/attendances.core.service.ts (hunk 3 above)
Comment (rushi-tekdi) [reply src/src/attendances/attendances.core.service.ts:246]: console log on dev is required for testing and track cron job flow. after testing complete on dev server then i will remove all logs and sent PR again

File: src/src/cron/prepareCertificateHtml.cron.ts (hunk 4 above)
Comment (rushi-tekdi) [reply src/src/cron/prepareCertificateHtml.cron.ts:38]: console log on dev is required for testing and track cron job flow. after testing complete on dev server then i will remove all logs and sent PR again

File: src/src/cron/prepareCertificateHtml.cron.ts (hunk 5 above)
Comment (rushi-tekdi) [reply src/src/cron/prepareCertificateHtml.cron.ts:79]: console log on dev is required for testing and track cron job flow. after testing complete on dev server then i will remove all logs and sent PR again

File: src/src/cron/prepareCertificateHtml.cron.ts (hunk 6 above)
Comment (rushi-tekdi) [reply src/src/cron/prepareCertificateHtml.cron.ts]: 2 is hours limit for last_processed. just like 4 we used in attedance face rekognition.
after testing done on dev I will move it in env variba

**Comment by manojLondhe** [comment]
@coderabbitai review

File: src/src/cron/faFaceIndexing.cron.ts (hunk 2 above)
Comment (rushi-tekdi) [reply src/src/cron/faFaceIndexing.cron.ts]: fixed this

File: src/src/attendances/attendances.core.service.ts (hunk 1 above)
Comment (rushi-tekdi) [reply src/src/attendances/attendances.core.service.ts:248]: @manojLondhe  our previous face rekognition 3 Cron Job contains console.log to debug issues.

**Pull request merged**